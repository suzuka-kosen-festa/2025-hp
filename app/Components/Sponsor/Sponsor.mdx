import { Canvas, Meta, ArgTypes, Source } from '@storybook/blocks';

import * as SponsorStories from "./Sponsor.stories";

<Meta of={SponsorStories} />

# Sponsorコンポーネント

Sponsorコンポーネントは、高専祭のスポンサー企業を階層的にレイアウトして表示するためのコンポーネントです。
スポンサーのサイズ（large、medium、small）に応じて異なるレイアウトと表示形式を提供します。

## 特徴

- **階層的表示**: スポンサーをサイズ別に分類して表示
- **多言語対応ソート**: 日本語の文字（ひらがな、カタカナ、漢字、英数字、記号）を適切にソート
- **画像とテキストの自動切り替え**: 有効な画像がある場合は画像を、ない場合はテキストを表示
- **レスポンシブデザイン**: 各サイズに最適化されたレイアウト
- **エラーハンドリング**: 画像読み込みエラー時の自動フォールバック

## インターフェース

### SponsorItem

```typescript
interface SponsorItem {
  name: string;           // スポンサー名
  image: string;          // 画像URL（空文字の場合はテキスト表示）
  size: "large" | "medium" | "small"; // スポンサーのサイズ
}
```

### SponsorComponentProps

```typescript
interface SponsorComponentProps {
  sponsors: SponsorItem[]; // 表示するスポンサーの配列
}
```

## 使用例

### 基本的な使用方法

<Canvas of={SponsorStories.Default} />

### サイズ別表示

各サイズの表示確認ができます。

<Canvas of={SponsorStories.SizeVariations} />

## 関数ロジックテスト

### isValidImage関数のテスト

画像URLの有効性判定ロジックをテストできます。有効なURL、空文字、null/undefined文字列、スペースのみのパターンを確認できます。

<Canvas of={SponsorStories.ImageValidationTest} />

### sortByJapaneseName関数のテスト

日本語文字列のソート機能をテストできます。ひらがな、カタカナ、英数字、記号の混在でのソート順序を確認できます。

期待されるソート順序：
`2年システム < 3D Innovation < 10年システム < B&K株式会社 < あいうえお株式会社 < カキクケコ株式会社 < ナニヌネノ株式会社`

<Canvas of={SponsorStories.JapaneseSortTest} />

## 設定とレイアウト

### サイズ別設定

<table>
  <thead>
    <tr>
      <th>サイズ</th>
      <th>Typography</th>
      <th>最大幅</th>
      <th>最大高さ</th>
      <th>レイアウト方向</th>
      <th>余白</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Large</td>
      <td>h4</td>
      <td>70%</td>
      <td>200px</td>
      <td>縦方向</td>
      <td>大</td>
    </tr>
    <tr>
      <td>Medium</td>
      <td>h5</td>
      <td>40%</td>
      <td>150px</td>
      <td>横方向</td>
      <td>中</td>
    </tr>
    <tr>
      <td>Small</td>
      <td>h6</td>
      <td>30%</td>
      <td>100px</td>
      <td>横方向</td>
      <td>小</td>
    </tr>
  </tbody>
</table>

### ソート機能

`sortByJapaneseName`関数は`Intl.Collator`を使用して日本語に最適化されたソートを実行します：

- **数値ソート**: 「10」は「2」より後に表示（`numeric: true`）
- **文字種混在対応**: ひらがな、カタカナ、漢字、英数字、記号の混在文字列も適切にソート
- **大文字優先**: 英語の大文字を小文字より先に表示（`caseFirst: "upper"`）
- **アクセント記号考慮**: より正確な文字比較（`sensitivity: "accent"`）
- **句読点考慮**: 句読点も含めた正確なソート（`ignorePunctuation: false`）

<Source code={`
function sortByJapaneseName(sponsors: SponsorItem[]): SponsorItem[] {
  return [...sponsors].sort((a, b) => {
    const options: Intl.CollatorOptions = {
      numeric: true, // 数値を数値として扱う（"10" > "2"）
      caseFirst: "upper", // 大文字を先に
      sensitivity: "accent", // アクセント記号を区別
      ignorePunctuation: false, // 句読点を考慮
      usage: "sort", // ソート用途で最適化
    };

    const collator = new Intl.Collator("ja-JP", options);
    return collator.compare(a.name, b.name);
  });
}
`} language="typescript" />

### 画像処理

`isValidImage`関数により以下の処理を行います：

1. **画像有効性チェック**: 
   - 空文字（`""`）
   - null、undefined 
   - "null"、"undefined"文字列
   - スペースのみの文字列（`"   "`）
   
   これらを無効として判定します

2. **エラーハンドリング**: 画像読み込み失敗時は自動的にテキスト表示にフォールバック
3. **レスポンシブ対応**: `object-fit: contain`で画像のアスペクト比を維持

<Source code={`
function isValidImage(imageUrl: string): boolean {
  return (
    imageUrl != null
    && imageUrl.trim() !== ""
    && imageUrl !== "undefined"
    && imageUrl !== "null"
  );
}
`} language="typescript" />

## Props

<ArgTypes of={SponsorStories.Default} />

## ベストプラクティス

### データ準備
```typescript
const sponsors: SponsorItem[] = [
  {
    name: "大手スポンサー株式会社",
    image: "/images/sponsors/large-sponsor.png",
    size: "large"
  },
  {
    name: "中堅企業",
    image: "", // 画像がない場合は空文字
    size: "medium"
  }
];
```

### 画像URL の指定
- 相対パス: `public/logo.png`
- 絶対パス: `/images/sponsors/logo.png`
- 外部URL: `https://example.com/logo.png`

### アクセシビリティ
- 画像には適切な`alt`属性が自動設定されます
- テキスト表示時は適切なヘディングレベルが使用されます
- セマンティックなHTML構造で実装されています

## トラブルシューティング

### 画像が表示されない場合
1. **画像URLが正しいか確認**
   - 相対パス: `public/logo.png`
   - 絶対パス: `/images/logo.png`
2. **画像ファイルが存在するか確認**
3. **ネットワークタブでHTTPエラーを確認**
4. **`isValidImage`関数のロジックを確認**
   - `ImageValidationTest` Storyで動作確認

### ソート順が期待通りでない場合
- **`JapaneseSortTest` Storyで確認**：期待されるソート順序と実際の表示を比較
- **日本語のソートは複雑**：特殊な文字が含まれる場合は個別に確認
- **数字を含む文字列**：数値として正しく比較されます（例：「10」>「2」）
- **`Intl.Collator`の設定**：ロケールが`ja-JP`になっているか確認

### 関数ロジックのデバッグ
1. **`ImageValidationTest`** を使用して`isValidImage`関数の動作を確認
2. **`JapaneseSortTest`** を使用して`sortByJapaneseName`関数の動作を確認
3. **ブラウザの開発者ツール**でコンソールログを確認