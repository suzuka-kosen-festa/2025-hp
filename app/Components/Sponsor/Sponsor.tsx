import type { ReactNode } from "react";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import React, { useState } from "react";

interface SponsorItem {
  name: string
  image: string
  size: "large" | "medium" | "small"
}

interface SponsorComponentProps {
  sponsors: SponsorItem[]
}

// 画像が有効かどうかをチェックする関数
function isValidImage(imageUrl: string): boolean {
  return Boolean(imageUrl && imageUrl.trim() !== "" && imageUrl !== "undefined" && imageUrl !== "null");
}

// サイズごとのvariantとfontSizeを一元管理
const SPONSOR_STYLE = {
  large: {
    variant: "h4" as const,
    fontSize: "1.5rem",
  },
  medium: {
    variant: "h5" as const,
    fontSize: "1.25rem",
  },
  small: {
    variant: "h6" as const,
    fontSize: "0.875rem",
  },
};

function Sponsor_Large({ name, image }: SponsorItem): ReactNode {
  const hasValidImage = isValidImage(image);
  const [imgError, setImgError] = useState(false);
  const style = SPONSOR_STYLE.large;

  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        margin: 1,
        padding: 2,
        maxWidth: "70%",
        textAlign: "center",
      }}
    >
      {hasValidImage && !imgError
        ? (
            <img
              src={image}
              alt={name}
              style={{
                maxWidth: "100%",
                maxHeight: "200px",
                objectFit: "contain",
              }}
              onError={() => setImgError(true)}
            />
          )
        : (
            <Typography
              variant={style.variant}
              component="div"
              sx={{
                fontWeight: "bold",
                color: "#333",
                lineHeight: 1.2,
              }}
            >
              {name}
            </Typography>
          )}
    </Box>
  );
}

function Sponsor_Medium({ name, image }: SponsorItem): ReactNode {
  const hasValidImage = isValidImage(image);
  const [imgError, setImgError] = useState(false);
  const style = SPONSOR_STYLE.medium;

  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        margin: 1,
        padding: 1.5,
        textAlign: "center",
        maxWidth: "40%",
      }}
    >
      {hasValidImage && !imgError
        ? (
            <img
              src={image}
              alt={name}
              style={{
                maxWidth: "100%",
                maxHeight: "150px",
                objectFit: "contain",
              }}
              onError={() => setImgError(true)}
            />
          )
        : (
            <Typography
              variant={style.variant}
              component="div"
              sx={{
                fontWeight: "bold",
                color: "#333",
                lineHeight: 1.2,
              }}
            >
              {name}
            </Typography>
          )}
    </Box>
  );
}

function Sponsor_Small({ name, image }: SponsorItem): ReactNode {
  const hasValidImage = isValidImage(image);
  const [imgError, setImgError] = useState(false);
  const style = SPONSOR_STYLE.small;

  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        margin: 0.5,
        padding: 1,
        maxWidth: "30%",
        textAlign: "center",
      }}
    >
      {hasValidImage && !imgError
        ? (
            <img
              src={image}
              alt={name}
              style={{
                maxWidth: "100%",
                maxHeight: "100px",
                objectFit: "contain",
              }}
              onError={() => setImgError(true)}
            />
          )
        : (
            <Typography
              variant={style.variant}
              component="div"
              sx={{
                fontWeight: "bold",
                color: "#333",
                lineHeight: 1.2,
              }}
            >
              {name}
            </Typography>
          )}
    </Box>
  );
}

// 五十音順でソートする関数
function sortByJapaneseName(sponsors: SponsorItem[]): SponsorItem[] {
  return [...sponsors].sort((a, b) => a.name.localeCompare(b.name, "ja"));
}

function Sponsor(props: SponsorComponentProps): ReactNode {
  const { sponsors } = props;

  // サイズ別にグループ化
  const largeSponsors = sortByJapaneseName(sponsors.filter(s => s.size === "large"));
  const mediumSponsors = sortByJapaneseName(sponsors.filter(s => s.size === "medium"));
  const smallSponsors = sortByJapaneseName(sponsors.filter(s => s.size === "small"));

  return (
    <Box sx={{ mt: 10 }}>
      {/* Largeサイズのスポンサー */}
      {largeSponsors.length > 0 && (
        <Box sx={{ mb: 20 }}>
          <Box
            sx={{
              display: "flex",
              flexDirection: "column",
              alignItems: "center",
              width: "80%",
              margin: "0 auto",
            }}
          >
            {largeSponsors.map(sponsor => (
              <Sponsor_Large key={`large-${sponsor.name}`} {...sponsor} />
            ))}
          </Box>
        </Box>
      )}

      {/* Mediumサイズのスポンサー */}
      {mediumSponsors.length > 0 && (
        <Box sx={{ mb: 20 }}>
          <Box
            sx={{
              display: "flex",
              flexWrap: "wrap",
              justifyContent: "flex-start",
              width: "80%",
              margin: "0 auto",
              gap: 0,
            }}
          >
            {mediumSponsors.map(sponsor => (
              <Sponsor_Medium key={`medium-${sponsor.name}`} {...sponsor} />
            ))}
          </Box>
        </Box>
      )}

      {/* Smallサイズのスポンサー */}
      {smallSponsors.length > 0 && (
        <Box sx={{ mb: 20 }}>
          <Box
            sx={{
              display: "flex",
              flexWrap: "wrap",
              justifyContent: "flex-start",
              width: "80%",
              margin: "0 auto",
              gap: 0,
            }}
          >
            {smallSponsors.map(sponsor => (
              <Sponsor_Small key={`small-${sponsor.name}`} {...sponsor} />
            ))}
          </Box>
        </Box>
      )}
    </Box>
  );
}

export default Sponsor;
