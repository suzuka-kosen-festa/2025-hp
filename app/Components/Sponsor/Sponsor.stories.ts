import type { Meta, StoryObj } from "@storybook/react";
import Sponsor from "./Sponsor";

const meta: Meta<typeof Sponsor> = {
  title: "Sponsor",
  component: Sponsor,
  parameters: {
    layout: "fullscreen",
  },

};

export default meta;
type Story = StoryObj<typeof Sponsor>;

// 最小限のデモデータ
const minimalSponsors = [
  {
    name: "テストスポンサー大",
    image: "public/logo-dark.png",
    size: "large" as const,
  },
  {
    name: "テストスポンサー中",
    image: "",
    size: "medium" as const,
  },
  {
    name: "テストスポンサー小",
    image: "",
    size: "small" as const,
  },
];

// isValidImage関数テスト用データ
const imageValidationTestData = [
  {
    name: "有効なURL",
    image: "public/logo-dark.png",
    size: "medium" as const,
  },
  {
    name: "空文字",
    image: "",
    size: "medium" as const,
  },
  {
    name: "null文字列",
    image: "null",
    size: "medium" as const,
  },
  {
    name: "undefined文字列",
    image: "undefined",
    size: "medium" as const,
  },
  {
    name: "スペースのみ",
    image: "   ",
    size: "medium" as const,
  },
];

// sortByJapaneseName関数テスト用データ
const sortTestSponsors = [
  {
    name: "ナニヌネノ株式会社",
    image: "",
    size: "medium" as const,
  },
  {
    name: "かいうえお株式会社",
    image: "",
    size: "medium" as const,
  },
  {
    name: "3D Innovation",
    image: "",
    size: "medium" as const,
  },
  {
    name: "アキクケコ株式会社",
    image: "",
    size: "medium" as const,
  },
  {
    name: "B&K株式会社",
    image: "",
    size: "medium" as const,
  },
  {
    name: "10年システム",
    image: "",
    size: "medium" as const,
  },
  {
    name: "2年システム",
    image: "",
    size: "medium" as const,
  },
  {
    name: "あきくけこ株式会社",
    image: "",
    size: "medium" as const,
  },
];

// 基本的なUIストーリー
export const Default: Story = {
  args: {
    sponsors: minimalSponsors,
  },
};

// isValidImage関数のロジックテストストーリー
export const ImageValidationTest: Story = {
  args: {
    sponsors: imageValidationTestData,
  },
  parameters: {
    docs: {
      description: {
        story: "isValidImage関数のロジックテスト: 有効なURL、空文字、null/undefined文字列、スペースのみのパターンを確認",
      },
    },
  },
};

// sortByJapaneseName関数のロジックテストストーリー
export const JapaneseSortTest: Story = {
  args: {
    sponsors: sortTestSponsors,
  },
  parameters: {
    docs: {
      description: {
        story: "sortByJapaneseName関数のロジックテスト: ひらがな、カタカナ、英数字、記号の混在でのソート順序を確認。期待順序: 2年システム < 3D Innovation < 10年システム < B&K株式会社 < あいうえお株式会社 < カキクケコ株式会社 < ナニヌネノ株式会社",
      },
    },
  },
};

// サイズ別表示テスト
export const SizeVariations: Story = {
  args: {
    sponsors: [
      { name: "Largeスポンサー", image: "", size: "large" as const },
      { name: "Mediumスポンサー", image: "", size: "medium" as const },
      { name: "Smallスポンサー", image: "", size: "small" as const },
    ],
  },
};
