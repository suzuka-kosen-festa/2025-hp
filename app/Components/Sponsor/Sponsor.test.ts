// @vitest-environment jsdom

import { composeStories } from "@storybook/react";

import { expect, it } from "vitest";

import * as stories from "./Sponsor.stories";

const { Default } = composeStories(stories);
it("sponsor snapshot", async () => {
  await Default.run();
  expect(document.body.firstChild).toMatchSnapshot();
});
