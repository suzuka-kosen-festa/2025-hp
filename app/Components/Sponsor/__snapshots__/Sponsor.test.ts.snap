// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sponsor snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-ib9iq6"
  >
    <div
      class="MuiBox-root css-1hfnxps"
    >
      <div
        class="MuiBox-root css-441yeq"
      >
        <div
          class="MuiBox-root css-3ib8bu"
        >
          <img
            alt="テストスポンサー大"
            src="public/logo-dark.png"
            style="max-width: 100%; max-height: 200px; object-fit: contain;"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1hfnxps"
    >
      <div
        class="MuiBox-root css-wdiapt"
      >
        <div
          class="MuiBox-root css-1vsv8c1"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 css-nto7bj-MuiTypography-root"
          >
            テストスポンサー中
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1hfnxps"
    >
      <div
        class="MuiBox-root css-wdiapt"
      >
        <div
          class="MuiBox-root css-1rzgnc4"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 css-1e5h870-MuiTypography-root"
          >
            テストスポンサー小
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
